README.md
# DEqMS
DEqMS is a tool for quantitative proteomic analysis, developed by Yafeng Zhu @ Karolinska Institutet. Manuscript in preparation.

## Installation
```{r}
#try http:// if https:// URLs are not supported
source("https://bioconductor.org/biocLite.R")
biocLite("DEqMS")
```
## Introduction
DEqMS works on top of Limma. However, Limma assumes same prior variance for all genes, the function `spectraCounteBayes` in DEqMS package  is able to correct the biase of prior variance estimate for genes identified with different number of PSMs/peptides. It works in a similar way to the intensity-based hierarchical Bayes method (Maureen A. Sartor et al BMC Bioinformatics 2006).
Outputs of `spectraCounteBayes`:

object is augmented form of "fit" object from `eBayes` in Limma, with the additions being:

`sca.t`     - Spectra Count Adjusted posterior t-value

`sca.p`     - Spectra Count Adjusted posterior p-value

`sca.dfprior` - estimated prior degrees of freedom

`sca.priorvar`- estimated prior variance

`sca.postvar` - estimated posterior variance

`loess.model` - fitted model

## analyze TMT labelled dataset
### 1. load R packages
```{r}
library(DEqMS)
```

### 2. Read input data and generate count table.
Since the input data used in DEqMS is PSM or peptide level data, it is highly recommended to filter them based protein level 1% FDR. (Grouping PSMs or peptides usually generate larger list of protiens)
The first two columns in input table should be peptide sequence and protein/gene names, intensity values for different samples start from 3rd columns. It is important the input file is arranged in this way.

Here we analyzed a published protemoics dataset (TMT10plex labelling) in which A431 cells (human epidermoid carcinoma cell line) were treated with three different miRNA mimics (Zhou Y et al. Oncogene 2016). [Pubmed](https://www.ncbi.nlm.nih.gov/pubmed/27477696)

```{r}
download.file("http://lehtiolab.se/Supplementary_Files/PXD004163.zip",
destfile = "./PXD004163.zip",method = "auto")
unzip("./PXD004163.zip")
dat.psm = readRDS("PXD004163.rds")
dat.psm[dat.psm == 0] <- NA # convert 0 to NA
dat.psm = na.omit(dat.psm) # remove rows with NAs

dat.psm.log = dat.psm
dat.psm.log[,3:12] =  log2(dat.psm[,3:12])  # log2 transformation

psm.count.table = as.data.frame(table(dat.psm$gene)) # generate PSM count table
rownames(psm.count.table)=psm.count.table$Var1
```
### 3. Generate sample annotation table.
```{r}
cond = c("ctrl","miR191","miR372","miR519","ctrl",
"miR372","miR519","ctrl","miR191","miR372")

sampleTable <- data.frame(
row.names = colnames(dat.psm)[3:12],
cond = as.factor(cond)
)
```

### 4. Summarization and Normalization
Choose one of the following functions to summarize peptide data to protein level. (Recommend median sweeping method)

 `group_col` is the column number you want to group by, set 2 if genes/proteins are in second column.
`ref_col`  is the columns where reference samples are.

1. use median sweeping method. [D'Angelo G et al JPR 2017](https://www.ncbi.nlm.nih.gov/pubmed/28745510) , [Herbrich SM et al JPR 2013](https://www.ncbi.nlm.nih.gov/pubmed/23270375)
```{r}
# medianSweeping does equal median normalization in the end
data.gene.nm = medianSweeping(dat.psm.log,group_col = 2)
```

2. calculate relative ratio using control/reference channels as denominator and then summarize to protein level by the median of all PSMs/Peptides.
```{r}
dat.gene = medianSummary(dat.psm.log,group_col = 2, ref_col=c(3,7,10))
dat.gene.nm = equalMedianNormalization(dat.gene)
```

3. summarize using Tukey's median polish procedure
```{r}
dat.gene = medpolishSummary(dat.psm.log,group_col = 2)
dat.gene.nm = equalMedianNormalization(dat.gene)
```

4. use Factor Analysis for Robust Microarray Summarization (FARMS)
see [Hochreiter S et al Bioinformatic 2007](http://bioinformatics.oxfordjournals.org/cgi/content/abstract/22/8/943), [Zhang B et al MCP 2017](https://www.ncbi.nlm.nih.gov/pubmed/28302922)
```{r}
# input is psm raw intensity, not log transformed values

dat.gene = farmsSummary(dat.psm,group_col = 2)
dat.gene.nm = equalMedianNormalization(dat.gene)
```

### 5. Differential gene expression analysis
Use the dataframe `dat.gene.nm` from Step 4 Summarization and Normalization as the input.

Since the genes in this data frame are aggregated from the PSM table, which has no FDR contorl at protein/gene level.
it is recommended that the data frame `dat.gene.nm` is filtered according to a gene ID list with 1% FDR.

```{r}
##skip this if you just want to follow this tutorial
genelist = read.table("example_genetable_0.01fdr.txt",header=T,sep="\t")
data.gene.nm = dat.gene.nm[rownames(dat.gene.nm) %in% genelist$gene,]
```
contitnue to DEqMS analysis
```{r}
gene.matrix = as.matrix(dat.gene.nm)
design = model.matrix(~cond,sampleTable)

fit1 <- eBayes(lmFit(gene.matrix,design))
fit1$count <- psm.count.table[rownames(fit1$coefficients),2]  # add an attribute containing PSM/peptide count for each gene

##check the values in the vector fit1$count
##if min(fit1$count) return NA or 0, you should troubleshoot the error before you continue
min(fit1$count)
head(fit1$count)

fit2 = spectraCounteBayes(fit1)
```
### 6. plot the fitted prior variance
Check fitted relation between piror variance and peptide/PSMs count works as expected. It should look similar to the plot below. Red curve is fitted value for prior variance, y is log pooled variances calculated for each gene.
```{r}
plotFitCurve(fit2,type = "boxplot", main="TMT10 dataset PXD004163", xlab="PSM count)
```

![My image](https://github.com/yafeng/DEqMS/blob/master/image/PXD004163.png)

### 7. Output the results
```{r}
DEqMS.results = outputResult(fit2,coef_col=3)
write.table(DEqMS.results, "DEqMS.analysis.out.txt", quote=F,sep="\t",row.names = F)
head(DEqMS.results,n=5)
```
| logFC        | AveExpr      | t            | P.Value  | adj.P.Val   | B           | gene    | PSMcount | sca.t        | sca.P.Value | sca.adj.pval |
|--------------|--------------|--------------|----------|-------------|-------------|---------|----------|--------------|-------------|--------------|
| -1.192424423 | -0.093472789 | -18.24327059 | 3.33E-08 | 0.000156241 | 8.895452085 | ANKRD52 | 17       | -19.01031999 | 4.21E-10    | 3.42E-06     |
| -1.177468714 | -0.051976035 | -16.52824778 | 7.66E-08 | 0.000235167 | 8.286987366 | CROT    | 21       | -17.7744113  | 8.96E-10    | 3.42E-06     |
| -1.241322465 | 0.072630242  | -18.20504475 | 3.39E-08 | 0.000156241 | 8.882911735 | TGFBR2  | 8        | -17.43066408 | 1.12E-09    | 3.42E-06     |
| -0.78072293  | 0.007763848  | -13.13285085 | 5.22E-07 | 0.00096131  | 6.742716671 | PDCD4   | 40       | -14.34371369 | 9.75E-09    | 2.24E-05     |
| -0.7976368   | -0.000657979 | -14.41697245 | 2.41E-07 | 0.000553767 | 7.388343266 | PHLPP2  | 8        | -12.7927884  | 3.42E-08    | 6.08E-05     |

Column `logFC`, `AveExpr`, `t`, `P.Value`, `adj.P.Val`, `B` are original values generated from Limma.
Last three columns `sca.t`, `sca.P.Value` and `sca.adj.pval` are values produced from `spectraCounteBayes`, which takes into account the number of quantified spectra/peptides.

## analyze label free dataset
### 1. load R packages
```{r}
library(DEqMS)
```
### 2. Read input data and experimental design.
Here we analyze a published label-free dataset in which they did quantitative proteomic analysis to detect proteome changes in FOXP3-overexpressed gastric cancer (GC) cells. (Pan D. et al 2017 Sci Rep) [Pubmed](https://www.ncbi.nlm.nih.gov/pubmed/29089565). The data was searched by MaxQuant Software and the output file "peptides.txt" was used here. (The column "Leading razor protein" in peptides.txt table was extracted as Protein column here).
```{r}
pepTable = readRDS(system.file("int/extdata/","PXD007725.rds",package = "DEqMS"))
```

### 3. Filter peptides based on missing values (DEqMS requires minimum two observations in each condition)
```{r}
pepTable[pepTable==0] <- NA
pepTable$cond1_na_count  = apply(pepTable,1, function(x) sum(is.na(x[3:7])))
pepTable$cond2_na_count  = apply(pepTable,1, function(x) sum(is.na(x[8:12])))

#require missing values no more than 3 in each condition
df.pep.filter =pepTable[pepTable$cond1_na_count<3 & pepTable$cond2_na_count <3,1:12]
```

In our tests, imputing methods have negatively affects the statistical accuracy. Therefore, we don't impute missing values here.

### 4.  calculate ratio using control samples and then summarize to protein level by the median of all PSMs/Peptides.
```{r}
df.pep.log = df.pep.filter
df.pep.log[,3:12] = log2(df.pep.log[,3:12])

protein.df = medianSummary(df.pep.log,group_col = 2,ref_col =8:12)
protein.df.nm = equalMedianNormalization(protein.df)
```
### 5. Differential expression analysis
```{r}
protein.matrix = as.matrix(protein.df.nm)

pep.count.table = as.data.frame(table(protein.df.nm$Protein))
rownames(pep.count.table) = pep.count.table$Var1

cond = as.factor(exp_design$condition)

design = model.matrix(~0+cond) # fitting without intercept
colnames(design) = c("AF","ANC")

fit1 = lmFit(protein.matrix,design = design)
cont <- makeContrasts(AF-ANC, levels = design)
fit2 = contrasts.fit(fit1,contrasts = cont)
fit3 <- eBayes(fit2)

fit3$count = pep.count.table[rownames(fit3$coefficients),2]

#check the values in the vector fit3$count
#if min(fit3$count) return NA or 0, you should troubleshoot the error before you continue
min(fit3$count)
head(fit3$count)

fit4 = spectraCounteBayes(fit3)
```

### 6. plot the fitted prior variance
```{r}
plotFitCurve(fit4,type = "boxplot",main = "Label-free dataset PXD0007725",xlab="peptide count")
```
![My image](https://github.com/yafeng/DEqMS/blob/master/image/PXD007725.png)

### 7. Output the results
```{r}
AF.results = outputResult(fit4,coef_col = 1)
write.table(AF.results,"AF.DEqMS.results.txt",sep = "\t",row.names = F,quote=F)
```


inst/NEWS
\name{NEWS}
\title{News for Package DEqMS}

\section{Version 0.99.4 (2018-09-11)}{
 \itemize{
  \item Package prepared for Bioconductor submission
 }
}


R/DEqMS.R
spectraCounteBayes<-function(fit,fit.method="loess",coef_col) {
    
    ################################################
    #  The function was adapted from:
    #  Function for IBMT (Intensity-based Moderated 
    #  T-statistic) Written by Maureen Sartor
    #  University of Cincinnati, 2006
    ################################################
    logVAR<-log(fit$sigma^2)
    df<-fit$df.residual
    numgenes<-length(logVAR[df>0])
    df[df==0]<-NA
    eg<-logVAR-digamma(df/2)+log(df/2)
    names(fit$count) = rownames(fit$coefficients)
    output<-fit

    if (fit.method == "loess"){
        x=log2(fit$count)
        loess.model = loess(logVAR~x,span = 0.75)
        y.pred = predict(loess.model)
        output$loess.model = loess.model
    }else if (fit.method == "nls"){
        x=fit$count
        y=fit$sigma^2
        nls.model =  nls(y~a+b/x,start = (list(a=0.1,b=0.05)))
        y.pred = log(predict(nls.model))
        output$nls.model = nls.model
    }
    
    egpred<-y.pred-digamma(df/2)+log(df/2)
    
    myfct<- (eg-egpred)^2 - trigamma(df/2)
    
    mean.myfct<-mean(myfct,na.rm=TRUE)
    
    priordf<-vector()
    testd0<-vector()
    
    for (i in seq(1,numgenes*10)) {
        testd0[i]<-i/10
        priordf[i]= abs(mean.myfct-trigamma(testd0[i]/2))
        if (i>2) {
            if (priordf[i-2]<priordf[i-1]) { break }
        }
    }
    d0<-testd0[match(min(priordf),priordf)] # prior degree found
    
    s02<-exp(egpred + digamma(d0/2) - log(d0/2)) # calculate prior variance
    
    post.var<- (d0*s02 + df*fit$sigma^2)/(d0+df)
    post.df<-d0+df
    # sca.t and scc.p stands for spectra count adjusted t and p values.  
    sca.t<-as.matrix(fit$coefficients[,coef_col]/(fit$stdev.unscaled[,coef_col]
    *sqrt(post.var)))
    sca.p<-as.matrix(2*(1-pt(abs(sca.t),post.df)))
    
    output$sca.t<-sca.t
    output$sca.p<-sca.p
    output$sca.postvar<-post.var
    output$sca.priorvar<-s02
    output$sca.dfprior<-d0
    
    return (output)
}

outputResult <-function(fit,coef_col=1){
    results.table = limma::topTable(fit,coef = coef_col,n= Inf)
    
    results.table$gene = rownames(results.table)
    results.table$count = fit$count[results.table$gene]
    
    results.table$sca.t = fit$sca.t[results.table$gene,coef_col]
    results.table$sca.P.Value = fit$sca.p[results.table$gene,coef_col]
    results.table$sca.adj.pval = p.adjust(results.table$sca.P.Value,
    method = "BH")
    results.table = results.table[order(results.table$sca.P.Value), ]
}

plotFitCurve <- function (fit,fit.method="loess",type = "boxplot",
    xlab="",main="") {
    x = fit$count
    y = fit$sigma^2
    
    if (fit.method=="nls"){
        model = fit$nls.model
        
        if (type=="scatterplot"){
            plot(log2(x),log(y),ylab="log(Variance)",xlab="log2(count)",
                main= main)
            
            y.pred <- log(predict(model))
            k = order(x)
            lines(log2(x[k]),y.pred[k],col='red',lwd=3)
            
        }else if (type=="boxplot"){
            df.temp = data.frame(pep_count =x, variance = y )
            df.temp.filter = df.temp[df.temp$pep_count<21,]
            boxplot(log(variance)~pep_count,df.temp.filter, xlab=xlab,
                    ylab = "log(Variance)", main=main)
            
            y.pred <- log(predict(model,data.frame(x=seq(1,20))))
            lines(seq(1,20),y.pred,col='red',lwd=3)
        }else{stop("only scatterplot and boxplot are supported")}
        
    }else if (fit.method=="loess"){
        model = fit$loess.model
        if (type=="scatterplot"){
            plot(log2(x),log(y),ylab="log(Variance)",main= main)
            
            y.pred <- predict(model)
            k = order(x)
            lines(log2(x[k]),y.pred[k],col='red',lwd=3)
            
        }else if (type=="boxplot"){
            df.temp = data.frame(pep_count =x, variance = y )
            df.temp.filter = df.temp[df.temp$pep_count<21,]
            boxplot(log(variance)~pep_count,df.temp.filter, xlab=xlab,
                    ylab = "log(Variance)", main=main)
            
            y.pred <- predict(model,log2(seq(1,20)))
            lines(seq(1,20),y.pred,col='red',lwd=3)
        }else{stop("only scatterplot and boxplot are supported")}
    }  
}

peptideProfilePlot <- function(dat,col=2,gene){
    
    dat.sub = dat[dat[,col]==gene,]
    
    colnames(dat.sub)[1]="Peptide"
    sample_size = ncol(dat.sub)-2
    
    m = reshape2::melt(dat.sub) 
    m$PSM_id =  rep(seq(1,nrow(dat.sub)),sample_size)
    ggplot(m, aes(x=variable,y=value))+
        geom_point()+
        geom_line(aes(group=PSM_id,col=Peptide))+
        theme(axis.text.x = element_text(angle = 90,hjust = 1))+
        ggtitle(dat.sub[1,2])+
        xlab("samples")+
        ylab("log2 intensity")+
        theme(plot.title = element_text(hjust = 0.5))
}


medianSummary <- function(dat,group_col=2,ref_col) {
    dat.ratio = dat
    if (length(ref_col)>1){
        dat.ratio[,3:ncol(dat)] = dat.ratio[,3:ncol(dat)] - 
            rowMeans(dat.ratio[,ref_col],na.rm = TRUE)  
    }else {
        dat.ratio[,3:ncol(dat)] = dat.ratio[,3:ncol(dat)] - dat.ratio[,ref_col]
    }
    
    dat.summary = plyr::ddply(dat.ratio,colnames(dat)[group_col],
    function(x)matrixStats::colMedians(as.matrix(x[,3:ncol(dat)]),na.rm = TRUE))
    
    colnames(dat.summary)[2:ncol(dat.summary)]=colnames(dat)[3:ncol(dat)]
    dat.new = dat.summary[,-1]
    rownames(dat.new) = dat.summary[,1]
    return (dat.new)
}


medianPolish <- function (m) {
    dat = medpolish(m,trace.iter=FALSE)$col
    return (dat)
}

medpolishSummary <- function(dat,group_col=2) {
    dat.summary = plyr::ddply(dat,colnames(dat)[group_col],
    function(x) medianPolish(as.matrix(x[,3:ncol(dat)])))
    
    colnames(dat.summary)[2:ncol(dat.summary)]=colnames(dat)[3:ncol(dat)]
    dat.new = dat.summary[,-1]
    rownames(dat.new) = dat.summary[,1]
    return (dat.new)
}


equalMedianNormalization <- function(dat) {
    sizefactor = matrixStats::colMedians(as.matrix(dat),na.rm = TRUE)
    dat.nm = sweep(dat,2,sizefactor)
    return (dat.nm)
}

medianSweeping <- function(dat,group_col=2) {
    dat.ratio = dat
    dat.ratio[,3:ncol(dat)] = dat.ratio[,3:ncol(dat)] - 
        matrixStats::rowMedians(as.matrix(dat.ratio[,3:ncol(dat)]),na.rm = TRUE)
    dat.summary = plyr::ddply(dat.ratio,colnames(dat)[group_col],
    function(x)matrixStats::colMedians(as.matrix(x[,3:ncol(dat)]),na.rm = TRUE))
    
    colnames(dat.summary)[2:ncol(dat.summary)] = colnames(dat)[3:ncol(dat)]
    dat.new = dat.summary[,-1]
    rownames(dat.new) = dat.summary[,1]
    
    dat.nm = equalMedianNormalization(dat.new)
    return (dat.nm)
}

farmsMethod <- function(df){
    if (nrow(df)==1){
        dat = log2(as.matrix(df))
    }else {dat = farms::generateExprVal.method.farms(
        as.matrix(na.omit(df)))$expr}
    return (dat)
}

farmsSummary <- function(dat,group_col=2) {
    dat.log = plyr::ddply(dat,colnames(dat)[group_col],
    function(x) farmsMethod(x[,3:ncol(dat)]))
    
    colnames(dat.log)[2:ncol(dat.log)]=colnames(dat)[3:ncol(dat)]
    
    dat.ratio = dat.log
    dat.ratio[,2:ncol(dat.ratio)]= dat.log[,2:ncol(dat.log)] - 
    matrixStats::rowMedians(as.matrix(dat.log[,2:ncol(dat.log)]),na.rm = TRUE)
    
    dat.summary = dat.ratio[,-1]
    rownames(dat.summary) = dat.ratio[,1]
    
    return (dat.summary)
}


inst/scripts/make-metadata.R
## validate with `AnnotationHub::readMetadataFromCsv("DEqMS")`
## (above pkg directory)

main.data <- data.frame(
    Title = c("microRNA treated A431 cell proteomics data"),
    Description = c("High resolution isoelectric focusing LC-MS/MS data for A431 cells treated by three different microRNA mimics"),
    RDataPath = c("DEqMS/PXD004163.rds"),
    BiocVersion="3.8",
    SourceType="text",
    Genome = "hg38",
    SourceVersion = "b3046142103a2033da2137086b5afd36",
    Coordinate_1_based = TRUE,
    SourceUrl="ftp://ftp.pride.ebi.ac.uk/pride/data/archive/2016/06/PXD004163/Yan_miR_PSM_table.flatmzidtsv.txt",
    Species="Homo Sapiens",
    TaxonomyId="9606",
    DataProvider="Swedish BioMS infrastructure",
    Maintainer="Yafeng Zhu <yafeng.zhu@ki.se>",
    RDataClass="data.frame",
    DispatchClass="Rds",
    stringsAsFactors = FALSE
)

write.csv(file="metadata.csv", main.data, row.names=FALSE)


inst/scripts/make-data.R
# Downloading the data.

url1 <- "ftp://ftp.pride.ebi.ac.uk/pride/data/archive/2016/06/PXD004163/Yan_miR_PSM_table.flatmzidtsv.txt"
download.file(url1, destfile = "./miR_PSMtable.txt",method = "auto")

url2 <- "ftp://ftp.pride.ebi.ac.uk/pride/data/archive/2016/06/PXD004163/Yan_miR_Protein_table.flatprottable.txt"
download.file(url2, destfile = "./miR_Proteintable.txt",method = "auto")


df.psm = read.table("miR_PSMtable.txt",stringsAsFactors = FALSE,header = TRUE,
                    quote = "", comment.char = "",sep = "\t")

df.prot = read.table("miR_Proteintable.txt",stringsAsFactors = FALSE,
                     header = TRUE, quote = "", comment.char = "",sep = "\t")

colnames(df.psm)

# filter at 1% peptide FDR
df.psm.sub = df.psm[df.psm$peptide.q.value<0.01,c(12,18,31:40)]

# filter at 1% protein FDR
df.psm.filter = df.psm.sub[df.psm.sub$HGNC.Gene.Symbol %in% df.prot$Protein.accession,]

# remove PSMs with missing values
df.psm.filter[df.psm.filter==0] <-NA 
df.psm.filter = na.omit(df.psm.filter)

colnames(df.psm.filter )[2]  = "gene"

saveRDS(df.psm.filter, "PXD004163.rds")